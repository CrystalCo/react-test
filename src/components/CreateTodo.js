import React, { Component, PropTypes } from 'react'
import moment from 'moment'
import { Modal, Header, Actions } from 'semantic-ui-react'
import {connect} from 'react-redux'
import {createTodo} from '../redux/modules/todo/actions'


export class CreateTodo extends Component {
  constructor() {
    super();
    this.state = {
      todo: {
        title: null,
        project: null
      },
      modalOpen: false
    }

    this.handleOpen=this.handleOpen.bind(this);
    this.handleClose=this.handleClose.bind(this);
    this.handleFieldChange=this.handleFieldChange.bind(this);
    this.handleCreate=this.handleCreate.bind(this);
  }

  handleOpen() {
    this.setState({modalOpen: true});
  }

  handleClose() {
    this.setState({todo: {title: null, project: null}});
    this.setState({modalOpen: false});
  }

  handleFieldChange(e) {
    var field = e.target.name;
    var value = e.target.value;
    var todo = this.state.todo;
    todo = Object.assign({}, todo, {[field]: value});
    this.setState({todo: todo});
  }

  handleCreate() {
    this.props.createTodo(this.state.todo);
    this.handleClose();
  }

  render() {
    return (
      <div className="column" style={{textAlign: 'center'}}>
        <Modal
          trigger={
            <div className="ui icon button" data-tooltip="Create New Todo" data-position="bottom center" onClick={this.handleOpen}>
              <i className="add icon"></i>
            </div>
          }
          open={this.state.modalOpen}
          onClose={this.handleClose}
        >
          <Modal.Header>Create Todo</Modal.Header>
          <Modal.Content>
            <div className='ui form'>
              <div className='field'>
                <label>Title</label >
                <input type='text' name="title" onChange={this.handleFieldChange} />
              </div>
              <div className='field'>
                <label>Project</label>
                <input type='text' name="project" onChange={this.handleFieldChange} />
              </div>
            </div>
          </Modal.Content>
          <Modal.Actions>
            <button className='ui inverted green button' onClick={this.handleCreate}>
              <i className='checkmark icon'></i> Create
            </button>
            <button className='ui inverted red button' onClick={this.handleClose}>
              <i className='remove icon'></i> Cancel
            </button>
          </Modal.Actions>
        </Modal>
      </div>
    )
  }
}

CreateTodo.propTypes = {
  createTodo: PropTypes.func.isRequired,
};

export default connect(null, {createTodo})(CreateTodo);