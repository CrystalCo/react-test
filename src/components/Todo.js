import React, { Component, PropTypes } from 'react'
import moment from 'moment'
import { Modal, Header, Actions } from 'semantic-ui-react'
import {connect} from 'react-redux'
import {editTodo, toggleTodo, deleteTodo} from '../redux/modules/todo/actions'
// var pathToImages = require.context('../images', true);

export class Todo extends Component {
  constructor() {
    super();
    this.state = {
      modalOpen: false,
      todo: {}
    }
    this.handleOpen=this.handleOpen.bind(this);
    this.handleClose=this.handleClose.bind(this);
    this.handleFieldChange=this.handleFieldChange.bind(this);
    this.handleEdit=this.handleEdit.bind(this);
    this.handleDelete=this.handleDelete.bind(this);
    this.handleToggle=this.handleToggle.bind(this);
  }

  handleOpen(e) {
    this.setState({modalOpen: true});
  }

  handleClose(e) {
    this.setState({modalOpen: false});
    this.setState({todo: {}});
  }

  handleFieldChange(e) {
    var field = e.target.name;
    var value = e.target.value;
    var todo = this.state.todo;
    var todo = Object.assign({}, todo, {[field]: value});
    this.setState({todo: todo});
  }

  handleEdit(e) {
    this.props.editTodo(this.props.id, this.state.todo);
    this.handleClose();
  }

  handleDelete(e) {
    this.props.deleteTodo(this.props.id);
  }

  handleToggle(e) {
    this.props.toggleTodo(this.props.id, {done: !this.props.done})
  }

  render() {
    return (
      <div className="column">
        <div className="ui brown card">
          <img className="ui image" src={this.props.url} />
          <div className="content">
            <div className="header">{this.props.title}</div>
            <div className="meta">{this.props.project}</div>
            <div className="meta">Created {moment(this.props.createdAt).fromNow()}</div>
          </div>
          <div className="extra content">
            <div>
              <div className="ui toggle checkbox" style={{marginBottom: '10px'}}>
                <input type="checkbox" name="public" value="on" defaultChecked ={this.props.done} onChange={this.handleToggle}/>
                <label>Complete</label>
              </div>
              <div className="ui two buttons">
                <Modal
                  trigger={<button className='ui basic green button' onClick={this.handleOpen}>Edit</button>}
                  open={this.state.modalOpen}
                  onClose={this.handleClose}
                >
                  <Modal.Header>Edit Todo</Modal.Header>
                  <Modal.Content>
                    <div className='ui form'>
                      <div className='field'>
                        <label>Title</label>
                        <input type='text'
                          name="title"
                          defaultValue={this.props.title}
                          onChange={this.handleFieldChange}
                        />
                      </div>
                      <div className='field'>
                        <label>Project</label>
                        <input type='text'
                          name="project"
                          defaultValue={this.props.project}
                          onChange={this.handleFieldChange}
                        />
                      </div>
                    </div>
                  </Modal.Content>
                  <Modal.Actions>
                    <button className='ui inverted green button' onClick={this.handleEdit}>
                      <i className='checkmark icon'></i> Update
                    </button>
                    <button className='ui inverted red button' onClick={this.handleClose}>
                      <i className='remove icon'></i> Cancel
                    </button>
                  </Modal.Actions>
                </Modal>
                <button className="ui red basic button" onClick={this.handleDelete}>Delete</button>
              </div>
            </div>
          </div>
        </div>
      </div>
    )
  }
}

Todo.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  project: PropTypes.string.isRequired,
  done: PropTypes.bool.isRequired,
  url: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  editTodo: PropTypes.func.isRequired,
  toggleTodo: PropTypes.func.isRequired,
  deleteTodo: PropTypes.func.isRequired
};

export default connect(null, {editTodo, toggleTodo, deleteTodo})(Todo);
