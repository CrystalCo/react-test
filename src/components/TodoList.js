import React, { Component } from 'react'
import moment from 'moment'
import { Modal, Header, Actions } from 'semantic-ui-react'
import ReactDOM from 'react-dom'
import Todo from './Todo'
import CreateTodo from './CreateTodo'
import url from '../images/React-icon.svg'

class TodoList extends Component {
  constructor() {
    super();
    this.state = {
      todos: [
        {
          "id": "b7bb6cfa-6e1b-4895-8198-4d5e080e5c0e",
          "title": "TodoX",
          "project": "ProjectX",
          "done": true,
          "createdAt": "2017-03-02T12:26:54.584Z",
          "modifiedAt": "2017-03-02T12:50:16.717Z"
        },
        {
          "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
          "title": "Todo1",
          "project": "Project",
          "done": false,
          "createdAt": "2017-03-02T23:04:38.003Z",
          "modifiedAt": "2017-03-02T23:05:30.133Z"
        },
        {
          "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c7",
          "title": "Todo5",
          "project": "Project5",
          "done": true,
          "createdAt": "2017-03-02T23:04:38.003Z",
          "modifiedAt": "2017-03-02T23:05:30.133Z"
        },
        {
          "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c6",
          "title": "Todo6",
          "project": "Project6",
          "done": true,
          "createdAt": "2017-03-02T23:04:38.003Z",
          "modifiedAt": "2017-03-02T23:05:30.133Z"
        }
      ],
    }
    this.handleCreate = this.handleCreate.bind(this);
    this.handleEdit = this.handleEdit.bind(this);
    this.handleDelete = this.handleDelete.bind(this);
    this.handleToggle = this.handleToggle.bind(this);
  }


  handleCreate(todo) {
    var newTodo = Object.assign({}, todo, {done: false, createdAt: new Date(), modifiedAt: new Date(), id: (Math.floor(Math.random() * 10000000)).toString()});
    var todos = this.state.todos;
    todos.push(newTodo);
    this.setState({todos: todos})
  }

  handleEdit(id, _todo) {
    var todos = this.state.todos;
    var newTodos = todos.map(todo => {
      if (todo.id === id) {
        return Object.assign({}, todo, _todo)
      } else {
        return todo;
      }
    });
    this.setState({todos: newTodos});
  }

  handleDelete(id) {
    var todos = this.state.todos;
    var newTodos = todos.filter(todo => {
      return todo.id !== id;
    });
    this.setState({todos: newTodos});
  }

  handleToggle(id) {
    var todos = this.state.todos;
    var newTodos = todos.map(todo => {
      if (todo.id === id) {
        return Object.assign({}, todo, {done: !todo.done})
      } else {
        return todo;
      }
    });
    this.setState({todos: newTodos});
  }

  render() {
    const { activeItem } = this.state

    return (
      <div>
        <div className="ui inverted segment" style={{borderRadius: '0px'}}>
          <h1 className="ui grey inverted center aligned header">
            <img className="ui large centered image" src={url} />
            <div className="content">React Testing</div>
          </h1>
        </div>
        <div className="ui centered container one column grid">
          <CreateTodo
            handleCreate={this.handleCreate}
          />
        </div>
        <div className="ui centered container three column grid">
          {this.state.todos.map(todo => {
            return (
              <Todo key={todo.id}
                id={todo.id}
                title={todo.title}
                project={todo.project}
                done={todo.done}
                createdAt={todo.createdAt}
                handleEdit={this.handleEdit}
                handleDelete={this.handleDelete}
                handleToggle={this.handleToggle}
              />
            )
          })}
        </div>
      </div>
    )
  }
}


ReactDOM.render(<TodoList />, document.getElementById('content'));