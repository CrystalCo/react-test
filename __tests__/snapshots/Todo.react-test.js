// Link.react-test.js
import React from 'react';
import {Todo} from '../../src/components/Todo';
import renderer from 'react-test-renderer';
import { shallow, mount, render } from 'enzyme';
import to_json from 'enzyme-to-json';
import moment from 'moment';

it('Renders correctly', () => {
  const component = renderer.create(
    <Todo
      id = '1'
      title = 'Todo'
      project = 'Project'
      done = {false}
      url = 'https://www.testurl.com/images/image'
      createdAt = {moment().subtract(1, 'days').format()}
      editTodo = {(id, todo) => { console.log(id, JSON.stringify(todo))}}
      toggleTodo = {(id, bool) => { console.log(id, bool)}}
      deleteTodo = {(id) => { console.log(id)}}
    ></Todo>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // // manually trigger the callback
  // tree.props.onMouseEnter();
  // // re-rendering
  // tree = component.toJSON();
  // expect(tree).toMatchSnapshot();

  // // manually trigger the callback
  // tree.props.onMouseLeave();
  // // re-rendering
  // tree = component.toJSON();
  // expect(tree).toMatchSnapshot();
});

it('Clicking edit button brings up modal', () => {
  const wrapper = shallow(
    <Todo
      id = '1'
      title = 'Todo'
      project = 'Project'
      done = {false}
      url = 'https://www.testurl.com/images/image'
      createdAt = {moment().subtract(1, 'days').format()}
      editTodo = {(id, todo) => { console.log(id, JSON.stringify(todo))}}
      toggleTodo = {(id, bool) => { console.log(id, bool)}}
      deleteTodo = {(id) => { console.log(id)}}
    ></Todo>
    )
  wrapper.find('button.ui.green.button').simulate('click');

  // wrapper.find('#sortByName').prop('className')).toEqual('SelectedBTN');
 // });
  expect(to_json(wrapper)).toMatchSnapshot();
})