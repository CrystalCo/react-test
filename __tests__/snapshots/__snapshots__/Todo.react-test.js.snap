// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clicking edit button brings up modal 1`] = `
<div
  className="column"
>
  <div
    className="ui brown card"
  >
    <img
      className="ui image"
      src="https://www.testurl.com/images/image"
    />
    <div
      className="content"
    >
      <div
        className="header"
      >
        Todo
      </div>
      <div
        className="meta"
      >
        Project
      </div>
      <div
        className="meta"
      >
        Created 
        25 days ago
      </div>
    </div>
    <div
      className="extra content"
    >
      <div>
        <div
          className="ui toggle checkbox"
          style={
            Object {
              "marginBottom": "10px",
            }
          }
        >
          <input
            defaultChecked={false}
            name="public"
            onChange={[Function]}
            type="checkbox"
            value="on"
          />
          <label>
            Complete
          </label>
        </div>
        <div
          className="ui two buttons"
        >
          <Modal
            closeOnDimmerClick={true}
            closeOnDocumentClick={false}
            dimmer={true}
            onClose={[Function]}
            open={false}
            trigger={
              <button
                className="ui basic green button"
                onClick={[Function]}
              >
                Edit
              </button>
            }
          >
            <ModalHeader>
              Edit Todo
            </ModalHeader>
            <ModalContent>
              <div
                className="ui form"
              >
                <div
                  className="field"
                >
                  <label>
                    Title
                  </label>
                  <input
                    defaultValue="Todo"
                    name="title"
                    onChange={[Function]}
                    type="text"
                  />
                </div>
                <div
                  className="field"
                >
                  <label>
                    Project
                  </label>
                  <input
                    defaultValue="Project"
                    name="project"
                    onChange={[Function]}
                    type="text"
                  />
                </div>
              </div>
            </ModalContent>
            <ModalActions>
              <button
                className="ui inverted green button"
                onClick={[Function]}
              >
                <i
                  className="checkmark icon"
                />
                 Update
              </button>
              <button
                className="ui inverted red button"
                onClick={[Function]}
              >
                <i
                  className="remove icon"
                />
                 Cancel
              </button>
            </ModalActions>
          </Modal>
          <button
            className="ui red basic button"
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renders correctly 1`] = `
<div
  className="column"
>
  <div
    className="ui brown card"
  >
    <img
      className="ui image"
      src="https://www.testurl.com/images/image"
    />
    <div
      className="content"
    >
      <div
        className="header"
      >
        Todo
      </div>
      <div
        className="meta"
      >
        Project
      </div>
      <div
        className="meta"
      >
        Created 
        a day ago
      </div>
    </div>
    <div
      className="extra content"
    >
      <div>
        <div
          className="ui toggle checkbox"
          style={
            Object {
              "marginBottom": "10px",
            }
          }
        >
          <input
            defaultChecked={false}
            name="public"
            onChange={[Function]}
            type="checkbox"
            value="on"
          />
          <label>
            Complete
          </label>
        </div>
        <div
          className="ui two buttons"
        >
          <button
            className="ui basic green button"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            Edit
          </button>
          <button
            className="ui red basic button"
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
